### PR TITLE
Allow app to provide FirebaseOptions and FirebaseMessaging instances.

### PR DESCRIPTION
**Pull request by brandonpage**

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+     * Will make call to Firebase.initializeApp if it hasn't already taken place.
+     *
+     * @param context Context
+     * @return True - if Firebase is now initialized, False - if the operation could not be completed.
+     */
+    private fun initializeFirebaseIfNeeded(context: Context): Boolean {
+        // If an instance of PushNotificationInterface is not provided the app does not intent to use the feature.
+        val pnDelegate = SalesforceSDKManager.getInstance().pushNotificationReceiver ?: return false
+        val appName = getAppNameForFirebase(context)
+        val firebaseOptions = pnDelegate.supplyFirebaseOptions() ?: FirebaseOptions.fromResource(context)
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt:144]: This is the first important line.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+            /*
+             * Performs registration steps if it is a new account, or if the
+             * account hasn't been registered yet. Otherwise, performs
+             * re-registration at the SFDC endpoint, to keep it alive.
+             */
+            if (account != null && !isRegistered(context, account)) {
+                setInProgress(context, account)
+
+                val firebaseMessaging = SalesforceSDKManager.getInstance()
+                    .pushNotificationReceiver?.supplyFirebaseMessaging() ?: FirebaseMessaging.getInstance()
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt:88]: This is the second important line.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+    @JvmStatic
+    fun unregister(context: Context, account: UserAccount?, isLastAccount: Boolean) {
+        if (isRegistered(context, account)) {
+            setInProgress(context, account)
+
+            // Deletes InstanceID only if there are no other logged in accounts.
+            if (isLastAccount) {
+                val firebaseInitialized = initializeFirebaseIfNeeded(context)
+                if (firebaseInitialized) {
+                    FirebaseMessaging.getInstance().isAutoInitEnabled = false
```
Comment (smcnulty-sfdc) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt:124]: You'll need to use the supplied here as well.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+     * that just logged in.
+     *
+     * @param context Context.
+     * @param account User account.
+     */
+    @JvmStatic
+    fun register(context: Context, account: UserAccount?) {
+        val firebaseInitialized = initializeFirebaseIfNeeded(context)
+        if (firebaseInitialized) {
+            FirebaseMessaging.getInstance().isAutoInitEnabled = true
```
Comment (smcnulty-sfdc) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt:77]: FirebaseMessaging.getInstance() will throw an exception if called and the [DEFAULT] FirebaseApp has an invalid key or is null.

**Review by JohnsonEricAtSalesforce — APPROVED** [review]
⭐️ Looks good to me!

**Review by smcnulty-sfdc — APPROVED** [review]

**Review by wmathurin — COMMENTED** [review]
LGTM
Where is the undocumented API that we fear might go away at some point?

**Comment by brandonpage** [comment]
> LGTM Where is the undocumented API that we fear might go away at some point?

Not in our code.  I'll slack you.

**Pull request merged**